### ISSUE TITLE
A method is called but never triggered

### ISSUE THREAD

**Issue by deepj** [oracle/truffleruby#1580]
OK. I guess this issue is connected to more issues what I've seen with TruffleRuby. But I finally find easy way to reproduce it and the code is short.

To reproduce:

```
git clone https://github.com/seattlerb/path_expander.git
cd path_expander
ruby test/test_path_expander.rb --name test_process_args_at
```

Error:
```
Run options: --name test_process_args_at --seed 11652

# Running:

F

Finished in 1.810316s, 0.5524 runs/s, 0.5524 assertions/s.

  1) Failure:
TestPathExpander#test_process_args_at [test/test_path_expander.rb:102]:
--- expected
+++ actual
@@ -1 +1 @@
-[["test/test_path_expander.rb"], ["--seed", "24"]]
+[[], []]


1 runs, 1 assertions, 1 failures, 0 errors, 0 skips
```

[process_args](https://github.com/seattlerb/path_expander/blob/master/lib/path_expander.rb#L75) method is called but the its code is not triggered. This methods is called in [assert_process_args](https://github.com/seattlerb/path_expander/blob/master/test/test_path_expander.rb#L27) method.

If I add `puts expander.process_args` before `assert_equal [exp_files, exp_args], expander.process_args` in [assert_process_args](https://github.com/seattlerb/path_expander/blob/master/test/test_path_expander.rb#L27) method. The method is triggered twice.

```ruby
def assert_process_args exp_files, exp_args, *args
  expander.args.concat args
  expander.process_args
  assert_equal [exp_files, exp_args], expander.process_args
 end
```

In this case I got this output from the test 

```
Run options: --name test_process_args_at --seed 55598

# Running:

F

Finished in 1.593521s, 0.6275 runs/s, 0.6275 assertions/s.

  1) Failure:
TestPathExpander#test_process_args_at [test/test_path_expander.rb:102]:
--- expected
+++ actual
@@ -1 +1 @@
-[["test/test_path_expander.rb"], ["--seed", "24"]]
+[["test/test_path_expander.rb", "test/test_path_expander.rb"], ["--seed", "24", "--seed", "24"]]


1 runs, 1 assertions, 1 failures, 0 errors, 0 skips
```

As you can see, the result of `expander.process_args` is doubled in the assert.

**Comment by aardvark179** [comment]
I've taken a look and the actual bug is not that `process_args` isn't called correctly, but that `Array#each` only iterates over the initial elements of the array, and excludes the extra ones pushed to it.

**Comment by deepj** [comment]
@eregon @chrisseaton this one please to RC13. I guess if @aardvark179 is right this issue might be connected to problems with splat operator such in 
https://github.com/oracle/truffleruby/issues/1469 + maybe this 
https://github.com/oracle/truffleruby/issues/1571 is relative too

Back this issue I guess I’ve seen more relative failed specs/tests with missing items in arrays. I just don’t remember which ones. I’m on phone today and the most of day tomorrow. I’m short here and I try to help prioritize things tomorrow in late evening.

**Comment by chrisseaton** [comment]
We were thinking about this one today in the group.

**Comment by aardvark179** [comment]
The fix for array iteration has been merged, and I this test now works.

**Comment by deepj** [comment]
@aardvark179 thanks. Could you try, please, this if it's associated to https://github.com/oracle/truffleruby/issues/1469 as well?
